pkg/miniogw/multipart.go

```diff
@@ -0,0 +1,438 @@
+		delete(uploads.pending, upload.ID)
+		uploads.mu.Unlock()
+
+		upload.Fail(err)
+		return "", err
+	}
+
+	go func() {
+		// setting zero value means the object never expires
+		expTime := time.Time{}
+
+		tempContType := metadata["content-type"]
+		delete(metadata, "content-type")
+
+		//metadata serialized
+		serMetaInfo := objects.SerializableMeta{
+			ContentType: tempContType,
+			UserDefined: metadata,
+		}
+
+		result, err := objectStore.Put(ctx, paths.New(object), upload.Stream, serMetaInfo, expTime)
+
+		uploads.mu.Lock()
+		delete(uploads.pending, upload.ID)
+		uploads.mu.Unlock()
+
+		if err != nil {
+			upload.Fail(err)
+		} else {
+			upload.Complete(minio.ObjectInfo{
+				Name:        object,
+				Bucket:      bucket,
+				ModTime:     result.Modified,
+				Size:        result.Size,
+				ETag:        result.Checksum,
+				ContentType: result.ContentType,
+				UserDefined: result.UserDefined,
+			})
+		}
+	}()
+
+	return upload.ID, nil
+}
+
+func (s *storjObjects) PutObjectPart(ctx context.Context, bucket, object, uploadID string, partID int, data *hash.Reader) (info minio.PartInfo, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	uploads := s.storj.multipart
+
+	upload, err := uploads.Get(bucket, object, uploadID)
+	if err != nil {
+		return minio.PartInfo{}, err
+	}
+
+	part, err := upload.Stream.AddPart(partID, data)
+	if err != nil {
+		return minio.PartInfo{}, err
+	}
+
+	err = <-part.Done
+	if err != nil {
+		return minio.PartInfo{}, err
+	}
+
+	partInfo := minio.PartInfo{
+		PartNumber:   part.Number,
+		LastModified: time.Now(),
+		ETag:         data.SHA256HexString(),
+		Size:         atomic.LoadInt64(&part.Size),
+	}
+
+	upload.mu.Lock()
+	upload.completed = append(upload.completed, partInfo)
+	upload.mu.Unlock()
+
+	return partInfo, nil
+}
+
+func (s *storjObjects) AbortMultipartUpload(ctx context.Context, bucket, object, uploadID string) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	uploads := s.storj.multipart
+
+	upload, err := uploads.Remove(bucket, object, uploadID)
+	if err != nil {
+		return err
+	}
+
+	errAbort := Error.New("abort")
+	upload.Stream.Abort(errAbort)
+	r := <-upload.Done
+	if r.Error != errAbort {
+		return r.Error
+	}
+	return nil
+}
+
+func (s *storjObjects) CompleteMultipartUpload(ctx context.Context, bucket, object, uploadID string, uploadedParts []minio.CompletePart) (objInfo minio.ObjectInfo, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	uploads := s.storj.multipart
+	upload, err := uploads.Remove(bucket, object, uploadID)
+	if err != nil {
+		return minio.ObjectInfo{}, err
+	}
+
+	// notify stream that there aren't more parts coming
+	upload.Stream.Close()
+	// wait for completion
+	result := <-upload.Done
+	// return the final info
+	return result.Info, result.Error
+}
+
+// func (s *storjObjects) ListMultipartUploads(ctx context.Context, bucket, prefix, keyMarker, uploadIDMarker, delimiter string, maxUploads int) (result minio.ListMultipartsInfo, err error) {
+// 	return
+// }
```

Comment:
Should the above commented code be removed if we don't need it for now?

Reply:
I don't feel strongly either way. I left them here more as a TODO.

Reply:
marked them more clearly as TODO, and moved them together